lib/Optimizer/Lower.cpp

```diff
@@ -647,6 +647,93 @@ static void lowerGroupConvolutionNode(Function *F, CompilationContext &cctx,
   replaceAllUsesOfWith(cctx.loweredInfoMap, BNG.getResult(), result);
 }
 
+static void lowerChannelwiseQuantizedConvolutionNode(
+    Function *F, CompilationContext &cctx,
+    const ChannelwiseQuantizedConvolutionNode &CQC) {
+  // ChannelwiseQuantizedConvolutionNode can be represented as a Concatenation
+  // of smaller dimension quantized Convolutions each with their own qparams.
+  // Input channels will be divided into equal groups of consecutive channels.
+  // These will be separately convolved each with its own filter and bias.
+  // This will result in 4 * Group + 1 nodes.
+
+  // Only lowering of groupwise, not channelwise is supported so far.
+  if (!CQC.getGroupwise()) {
+    return;
+  }
+
+  llvm::ArrayRef<unsigned_t> kernels = CQC.getKernels();
+  llvm::ArrayRef<unsigned_t> pads = CQC.getPads();
+  llvm::ArrayRef<unsigned_t> strides = CQC.getStrides();
+  unsigned_t group = CQC.getGroup();
+  auto in = CQC.getInput();
+
+  Constant *filter = llvm::cast<Constant>(CQC.getFilter());
+  Constant *bias = llvm::cast<Constant>(CQC.getBias());
+  Constant *scales = llvm::cast<Constant>(CQC.getScales());
+  Constant *offsets = llvm::cast<Constant>(CQC.getOffsets());
+
+  ShapeNHWC idim = ShapeNHWC(in.dims());
+  ShapeHW kdim(kernels);
+  unsigned inCperG = idim.c / group;
+  unsigned outCperG = filter->dims()[0] / group;
+
+  auto convOutDims = CQC.getResult().dims().vec();
+  convOutDims[3] = outCperG;
+
+  auto filterDims = filter->dims().vec();
+  filterDims[0] = outCperG;
+  filterDims[3] = inCperG;
+
+  // Final output type of each convolution after rescaling
+  auto finalConvOutTy = F->getParent()->uniqueTypeWithNewShape(
+      CQC.getResult().getType(), convOutDims);
+
+  auto scalesHandle = scales->getHandle<float>();
+  auto offsetsHandle = offsets->getHandle<int32_t>();
+
+  Module *M = F->getParent();
+
+  std::vector<NodeValue> branches;
+  for (unsigned_t groupId = 0; groupId < group; groupId++) {
+    float filterScale = scalesHandle.raw(groupId);
+    int32_t filterOffset = offsetsHandle.raw(groupId);
+
+    SliceNode *inSlice =
+        F->createSlice(CQC.getName(), in, {0, 0, 0, groupId * inCperG},
+                       {idim.n, idim.h, idim.w, (groupId + 1) * inCperG});
+
+    // Create quantized filter sliced Constant.
+    Tensor slicedFilterTensor = filter->getPayload().getOwnedSlice(
```

Comment:
I don't think we need an owned slice here or below since you're then copying over the slice anyway into the Constant via `createConstant()`, right?

Reply:
Ok I std::moved the tensors to the constants now so there's no additional copy. I tried using unowned tensors, put them into a constant, get mutable payload reference then modify that tensor and it looks more confusing than it's worth imo